### ISSUE TITLE
Type inference failure with function types with type arguments

### ISSUE THREAD

**Issue by abitofevrything** [dart-lang/sdk#51590]
The easiest way to explain this is with some code:
```dart
R callWithSomeTypeArgument<R>(R Function<T>() function) => function<String>();

void main() {
  // The argument type 'List<E> Function<E>({bool growable})' can't
  // be assigned to the parameter type 'List<Z0> Function<T>()'.
  final test = callWithSomeTypeArgument(List.empty);
  dynamic test2 = callWithSomeTypeArgument(List.empty);
  
  // All OK!
  List<dynamic> test3 = callWithSomeTypeArgument(List.empty);
}
```

This should work - I think - with the type of `test` being inferred as `List<dynamic>` since we have no information on what `T` will be. Instead it fails and the error message contains a reference to some `Z0` type which is I assume generated for type inference.

**Comment by lrhn** [comment]
@stereotype441 
Is this working as intended? 

I'm guessing we treat no type and `dynamic` as imposing no context type, so `R` has to be found by upwards inference. We then try to solve `List<E> Function<E>({...}) <: R Function<T>()` for `R` and fail because we don't just want to guess `Object?` as a solution, and there is no `List<X>` that works.

**Comment by abitofevrything** [comment]
Doesn't List<dynamic> work here? We know the function returns a List because the return type is List<E> but we can't guess E. I feel that guessing dynamic for E is then ok in this case.

**Comment by lrhn** [comment]
True, `List<Object?>` is also a solution. So maybe it's because the inference decides on `List<T>`, then figures out that it doesn't actually know `T` and gives up? (But now I'm just guessing.)

**Comment by abitofevrything** [comment]
After some more testing, I have found out that this issue only occurs in the analyzer. The frontend correctly infers the type argument and reports no errors, so commands like `dart compile` work fine - `dart analyze` continues to report the error though.

**Comment by abitofevrything** [comment]
A bit more investigation and the issue seems to arise [here](https://github.com/dart-lang/sdk/blob/cfa03ed97198307f265dfd8071190164eeb4bc81/pkg/analyzer/lib/src/dart/element/type_constraint_gatherer.dart#L427). This line incorrectly gathers that `R` should be a supertype of `List<Z0>` when it should gather that `R` is a supertype of `List<E>`, as it substitues `Z0` for `E` in the check.

I'm not sure if this is a bug in the analyzer implementation or an issue in the spec itself.
